src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,130 @@
+ *      (e.g. rotating device id)
+ *
+ */
+
+#include "AdditionalDataPayloadGenerator.h"
+#include "AdditionalDataPayload.h"
+
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <crypto/CHIPCryptoPAL.h>
+#include <stdlib.h>
+#include <support/CHIPMem.h>
+
+using namespace chip;
+using namespace chip::System;
+using namespace chip::TLV;
+using namespace chip::Crypto;
+using namespace chip::SetupPayload;
+
+AdditionalDataPayloadGenerator::AdditionalDataPayloadGenerator(uint16_t rotationCounter, char * serialNumberBuffer,
+                                                               size_t serialNumberBufferSize) :
+    mRotationCounter(rotationCounter),
+    mSerialNumberBuffer(serialNumberBuffer), mSerialNumberBufferSize(serialNumberBufferSize)
+{
+    mHashInput =
+        reinterpret_cast<char *>(chip::Platform::MemoryAlloc(ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + serialNumberBufferSize));
+}
+
+AdditionalDataPayloadGenerator::~AdditionalDataPayloadGenerator()
+{
+    if (mHashInput != nullptr)
+    {
+        chip::Platform::MemoryFree(mHashInput);
+        mHashInput = nullptr;
+    }
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(PacketBufferHandle & bufferHandle)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_LENGTH + 1];
+    uint8_t rotatingDeviceIdBufferSize;
+
+    // Initialize TLVWriter
+    writer.Init(PacketBuffer::New());
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    // Generating Device Rotating Id
+    err = generateRotatingDeviceId(rotatingDeviceIdBuffer, rotatingDeviceIdBufferSize);
+    SuccessOrExit(err);
+
+    // Adding the rotating device id to the TLV data
+    err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+    SuccessOrExit(err);
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(char * rotatingDeviceIdBuffer,
+                                                                    uint8_t & rotatingDeviceIdBufferSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    char rotationCounterStr[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH];
+
+    char output[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    uint8_t outputIndex     = 0;
+    uint8_t hashOutputIndex = 0;
+
+    sprintf(rotationCounterStr, "%d", mRotationCounter);
+    strcpy(mHashInput, rotationCounterStr);
+    strcat(mHashInput, mSerialNumberBuffer);
+
+    err = Hash_SHA256(reinterpret_cast<unsigned char *>(mHashInput),
+                      ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize, hashOutputBuffer);
```

Comment:
If the rotationCounterStr did not use the entire width of ROTATING_DEVICE_ID_COUNTER_STR_LENGTH, this will hash garbage data at the end of the buffer. Again, please use fixed-length binary encoding altogether.

Reply:
Fixed by padding the string